Add support for the tracecontext standard

bram123 opened the pull request:
- new traceMode config option to switch between traceid and tracecontext
  - supporting both options at the same time is out of scope for this PR, might be possible in the future
  - Need to think about what happens when X-TraceId header is in the request with a value that's not valid for traceContext standard, but app is doing ajax calls with traceparent headers.

bram123 committed:
Add support for the tracecontext standard
- new traceMode config option to switch between traceid and tracecontext
-- supporting both options at the same time is out of scope for this PR, might be possible in the future
-- Need to think about what happens when X-TraceId header is in the request with a value that's not valid for traceContext standard, but app is doing ajax calls with traceparent headers.

bram123 committed:
Fix phpstan issues

bram123 committed:
Fix phpstan issues

bram123 committed:
Fix phpstan issues

frankdekker commented on src/Service/TraceContextService.php:
```diff
@@ -0,0 +1,83 @@
+    public const HEADER_TRACEPARENT = 'traceparent';
+    public const HEADER_TRACESTATE  = 'tracestate';
+
+    public function validateTraceParent(string $traceParent): bool
```
Change signature to `public function supports(Request $request): bool`.

frankdekker commented on src/Service/TraceContextService.php:
```diff
@@ -0,0 +1,83 @@
+        return implode(',', $traceState);
+    }
+
+    public function parseTraceContext(string $traceParent, string $traceState): TraceContext
```
Move this method below the supports method and pass the Request object.

Pref: `public function getTraceContext(Request $request): TraceContext`

frankdekker commented on src/Http/TraceContextAwareHttpClient.php:
```diff
@@ -0,0 +1,67 @@
+
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        if ($this->storage->getTrace() instanceof TraceContext) {
```
Instead of having specific TraceContextService input here. Pass the `method`, `url`, and `options` to the service and let the service manipulate the options. This way we can swap in a different service and set different headers.

Suggestion:
`$options = $this->traceContextService->applyTraceHeaders(string $method, string $url, array $options): array`

Or something like

`$options['headers'] += $this->traceContextService->getTraceheaders(string $method, string $url, array $options): array`

frankdekker commented on src/Http/TraceIdAwareHttpClient.php:
```diff
@@ -15,9 +15,9 @@
 class TraceIdAwareHttpClient implements HttpClientInterface, ResetInterface, LoggerAwareInterface
```
This class would then be obsolete. We would just swap the TraceContextService.

bram123 committed:
Baseline phpmd notice

bram123 committed:
Baseline phpmd notice

bram123 committed:
Baseline phpmd notice

bram123 committed:
Add extra unittests

bram123 committed:
Merge remote-tracking branch 'origin/master' into add_tracecontext

# Conflicts:
#	README.md
#	phpmd.baseline.xml
#	phpstan-baseline.neon
#	src/DependencyInjection/Compiler/HttpClientTraceIdPass.php
#	src/DependencyInjection/SymfonyTraceExtension.php
#	src/EventSubscriber/CommandSubscriber.php
#	src/EventSubscriber/MessageBusSubscriber.php
#	src/EventSubscriber/TraceIdSubscriber.php
#	src/Generator/TraceId/RamseyUuid4Generator.php
#	src/Generator/TraceId/SymfonyUuid4Generator.php
#	src/Generator/TraceId/TraceIdGeneratorInterface.php
#	src/Http/TraceIdAwareHttpClient.php
#	src/Monolog/TraceIdProcessor.php
#	src/Twig/TraceIdExtension.php
#	tests/Functional/App/Service/TestIdStorage.php
#	tests/Functional/App/TestKernel.php
#	tests/Functional/App/config/config.yml
#	tests/Functional/ApplicationTest.php
#	tests/Functional/HttpClientTest.php
#	tests/Functional/MessengerTest.php
#	tests/Functional/RequestHandleTest.php
#	tests/Unit/EventSubscriber/CommandSubscriberTest.php
#	tests/Unit/EventSubscriber/MessageBusSubscriberTest.php
#	tests/Unit/EventSubscriber/TraceIdSubscriberTest.php
#	tests/Unit/Generator/TraceId/RamseyUuid4GeneratorTest.php
#	tests/Unit/Generator/TraceId/SymfonyUuid4GeneratorTest.php
#	tests/Unit/Http/TraceIdAwareHttpClientTest.php
#	tests/Unit/Monolog/TraceIdProcessorTest.php
#	tests/Unit/TraceStorageTest.php
#	tests/Unit/Twig/TraceIdExtensionTest.php

bram123 committed:
Update branch with master

bram123 committed:
Fix function test config files after merge

bram123 committed:
Merge subscribers/clients together, move difference in logic into separate Trace Service classes
- split TraceContextService parse logic into separate Parse class

bram123 committed:
Register TraceContextIdGenerator as TraceIdGeneratorInterface

bram123 committed:
Remove TraceId object, always use TraceContext

MessageBusSubscriber now passes the full TraceContext object to the TraceStamp
- clones original TraceContext, fills parentTransactionId in cloned trace

bram123 committed:
Fix phpstan warning
Update baseline

bram123 committed:
Rename occurences of traceId to trace

bram123 committed:
Update readme documentation for both traceid and tracecontext modes

frankdekker commented on README.md:
```diff
@@ -41,57 +34,9 @@ return [
+It's also possible to configure the bundle to setup custom request/response headers and custom ID generators.  
+Read more about the [TraceId configuration](docs/configuration/traceid.md) and [TraceContext configuration](docs/configuration/tracecontext.md) in the /docs pages.
 
 ## How it Works
```
Could you update this description below aswell? With focus on the TraceContext feature.

bram123 committed:
Update readme documentation

bram123 committed:
Update readme documentation

bram123 committed:
Update readme documentation
